torch/_refs/__init__.py

```diff
@@ -5599,12 +5599,16 @@ def eye(
     torch._check(n >= 0, lambda: f"n must be greater or equal to 0, got {n}")
     torch._check(m >= 0, lambda: f"m must be greater or equal to 0, got {m}")
 
-    range_n = torch.arange(n, dtype=torch.int64, device=device, requires_grad=False)
-    range_m = torch.arange(m, dtype=torch.int64, device=device, requires_grad=False)
+    if max(n, m) <= torch.iinfo(torch.int32).max:
+        range_dtype = torch.int32  # optimization for small tensors
```

Comment:
2 billon items is quite a lot :) so for small tensors, maybe also supporting uint16 (65536) would be better?

Reply:
True, though I'm mainly trying to avoid int64 math on CUDA (which is much slower).  While the difference between int16 and int32 is smaller.